File: app/src/main/java/io/github/hrilab/pepper_realtime/managers/QuizDialogManager.java
package io.github.hrilab.pepper_realtime.managers;

import android.app.AlertDialog;
import android.content.Context;
import android.content.res.ColorStateList;
import android.graphics.Color;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.TextView;

import androidx.core.content.ContextCompat;

import io.github.hrilab.pepper_realtime.ChatActivity;
import io.github.hrilab.pepper_realtime.R;

import java.util.ArrayList;
import java.util.List;

/**
 * Manager for Quiz Dialog UI functionality.
 * Handles the creation and management of quiz popups.
 */
public class QuizDialogManager {
    
    private static final String TAG = "QuizDialogManager";
    
    /**
     * Interface for quiz dialog callbacks
     */
    public interface QuizDialogCallback {
        void onQuizAnswered(String question, String selectedOption);
        void onInterruptRequested();
        boolean isActivityFinishing();
        boolean shouldInterrupt();

    }
    
    /**
     * Show a quiz dialog with the given question and options
     * 
     * @param context The activity context
     * @param question The quiz question
     * @param options Array of exactly 4 answer options
     * @param correctAnswer The correct answer from the options
     * @param callback Callback interface for handling quiz events
     */
    public static void showQuizDialog(Context context, String question, String[] options, 
                                    String correctAnswer, QuizDialogCallback callback) {
        
        if (callback.isActivityFinishing()) { 
            Log.w(TAG, "Not showing quiz dialog because activity is finishing.");
            return; 
        }
        
        ChatActivity activity = (ChatActivity) context;
        View dialogView = activity.getLayoutInflater().inflate(R.layout.dialog_quiz, null);
        AlertDialog.Builder builder = new AlertDialog.Builder(activity, R.style.QuizDialog);
        builder.setView(dialogView);
        builder.setCancelable(false);
        
        // Get references to the UI elements
        TextView questionTextView = dialogView.findViewById(R.id.quiz_question_textview);
        Button option1Button = dialogView.findViewById(R.id.quiz_option_1);
        Button option2Button = dialogView.findViewById(R.id.quiz_option_2);
        Button option3Button = dialogView.findViewById(R.id.quiz_option_3);
        Button option4Button = dialogView.findViewById(R.id.quiz_option_4);
        List<Button> buttons = new ArrayList<>();
        buttons.add(option1Button);
        buttons.add(option2Button);
        buttons.add(option3Button);
        buttons.add(option4Button);

        // Set question and button texts
        questionTextView.setText(question);
        for (int i = 0; i < buttons.size() && i < options.length; i++) {
            buttons.get(i).setText(options[i]);
        }

        final AlertDialog dialog = builder.create();

        // Set onClick listeners for each button
        for (final Button button : buttons) {
            button.setOnClickListener(v -> {
                // Interrupt robot if it's currently speaking
                if (callback.shouldInterrupt()) {
                    callback.onInterruptRequested();
                }

                String selectedOption = button.getText().toString();
                boolean isCorrect = selectedOption.equals(correctAnswer);

                // Disable all buttons to prevent multiple answers
                for (Button b : buttons) {
                    b.setEnabled(false);
                }

                // Color the buttons based on the answer
                if (isCorrect) {
                    int green = ContextCompat.getColor(context, R.color.correct_green);
                    button.setBackgroundTintList(ColorStateList.valueOf(green));
                    button.setTextColor(Color.WHITE);
                } else {
                    int red = ContextCompat.getColor(context, R.color.incorrect_red);
                    button.setBackgroundTintList(ColorStateList.valueOf(red));
                    button.setTextColor(Color.WHITE);
                    for (Button b : buttons) {
                        if (b.getText().toString().equals(correctAnswer)) {
                            int green = ContextCompat.getColor(context, R.color.correct_green);
                            b.setBackgroundTintList(ColorStateList.valueOf(green));
                            b.setTextColor(Color.WHITE);
                        }
                    }
                }

                // Notify callback about the selected answer
                callback.onQuizAnswered(question, selectedOption);

                // Close the dialog after a delay
                new Handler(Looper.getMainLooper()).postDelayed(dialog::dismiss, 4000); // 4 seconds delay
            });
        }
        
        dialog.show();
        // Make dialog full-screen
        if (dialog.getWindow() != null) {
            dialog.getWindow().setLayout(ViewGroup.LayoutParams.MATCH_PARENT, ViewGroup.LayoutParams.MATCH_PARENT);
        }
        
        Log.i(TAG, "Quiz dialog shown with question: " + question);
    }
}

File: app/src/main/java/io/github/anonymous/pepper_realtime/manager/QuizDialogManager.kt
package io.github.anonymous.pepper_realtime.manager

import android.util.Log
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.getValue
import androidx.compose.runtime.setValue

/**
 * Manager for Quiz Dialog UI functionality.
 * Handles the creation and management of quiz popups using Jetpack Compose.
 */
object QuizDialogManager {

    private const val TAG = "QuizDialogManager"

    /**
     * Interface for quiz dialog callbacks
     */
    interface QuizDialogCallback {
        fun onQuizAnswered(question: String, selectedOption: String)
        fun onInterruptRequested()
        fun isActivityFinishing(): Boolean
        fun shouldInterrupt(): Boolean
    }

    /**
     * State holder for the quiz dialog
     */
    data class QuizState(
        val isVisible: Boolean = false,
        val question: String = "",
        val options: List<String> = emptyList(),
        val correctAnswer: String = "",
        val callback: QuizDialogCallback? = null
    )

    // Observable state for Compose
    var quizState by mutableStateOf(QuizState())
        private set

    /**
     * Show a quiz dialog with the given question and options
     *
     * @param question The quiz question
     * @param options Array of exactly 4 answer options
     * @param correctAnswer The correct answer from the options
     * @param callback Callback interface for handling quiz events
     */
    fun showQuizDialog(
        question: String,
        options: Array<String>,
        correctAnswer: String,
        callback: QuizDialogCallback
    ) {
        if (callback.isActivityFinishing()) {
            Log.w(TAG, "Not showing quiz dialog because activity is finishing.")
            return
        }

        quizState = QuizState(
            isVisible = true,
            question = question,
            options = options.toList(),
            correctAnswer = correctAnswer,
            callback = callback
        )

        Log.i(TAG, "Quiz dialog shown with question: $question")
    }

    /**
     * Called when user selects an answer
     */
    fun onAnswerSelected(selectedOption: String) {
        val state = quizState
        val callback = state.callback ?: return

        // Interrupt robot if it's currently speaking
        if (callback.shouldInterrupt()) {
            callback.onInterruptRequested()
        }

        // Notify callback about the selected answer
        callback.onQuizAnswered(state.question, selectedOption)
    }

    /**
     * Dismiss the quiz dialog
     */
    fun dismissQuiz() {
        quizState = QuizState()
        Log.i(TAG, "Quiz dialog dismissed")
    }
}
